File: app/src/api/kalaidoscope/organize.ts
import type { Result } from "neverthrow";
import { withActiveClient } from "./_active";

/**
 * Workspace organise-pipeline status, mirroring `api.OrganizeStatus` from
 * `GET /api/organize`. Derived server-side from rows plus the workers' own
 * in-flight flags; nothing is stored, so re-fetch whenever the underlying
 * collections change (see `useOrganizeStatus`).
 */
export type MapState =
  | "empty"
  | "unannotated"
  | "annotating"
  | "consolidating"
  | "folding"
  | "settled";

export type DiscoverState =
  | "never_run"
  | "pending"
  | "running"
  | "due"
  | "settled";

export type DiscoverKind = "colours" | "projections" | "reflections";

export interface RunInfo {
  id: string;
  status: string;
  error?: string;
  model?: string;
  rounds?: number;
  mapVersion?: number;
  finished: string;
  /** The row says running but no worker in this process is running it. */
  interrupted?: boolean;
}

export interface OrganizeStatus {
  fragments: number;
  imports: { pending: number; lastError?: string };
  map: {
    state: MapState;
    version: number;
    annotated: number;
    pendingAnnotation: number;
    unconsolidated: number;
    lastRun?: RunInfo;
    lastDrainError?: string;
  };
  discover: {
    state: DiscoverState;
    running?: DiscoverKind | "";
    pending: DiscoverKind[];
    due: DiscoverKind[];
    runs: Partial<Record<DiscoverKind, RunInfo>>;
    proposals: { projections: number; reflections: number };
  };
  /** Waves start on their own (KALAIDO_AUTO_WAVE) or only from Start. */
  policy?: { wave: boolean };
  reconcile: ReconcileStatus;
  colour?: ColourStatus;
}

export interface ColourStatus {
  promptColoursCount: number;
  totalColoursCount: number;
  unjudgedFragments: number;
  draining: boolean;
}

export interface CurrentEntityInfo {
  id: string;
  type: string;
}

export interface WaveProgress {
  completed: number;
  total: number;
}

/**
 * The reconcile wave's state. `lastStarted` at or after a Start press, with
 * `running` false, means the wave that press began has ended.
 */
export interface ReconcileStatus {
  running: boolean;
  /** RFC3339; absent until a wave has started. */
  lastStarted?: string;
  /** What ended the most recent wave; absent when it ran clean. */
  lastError?: string;
  /** RFC3339; absent until a wave has completed without error. */
  lastCompleted?: string;
  /** RFC3339; absent unless the most recent wave was cancelled. */
  lastCancelled?: string;
  currentEntity?: CurrentEntityInfo;
  progress?: WaveProgress;
}

export async function getOrganizeStatus(): Promise<
  Result<OrganizeStatus, Error>
> {
  return withActiveClient((client) =>
    client.send<OrganizeStatus>("/api/status", { method: "GET" }),
  );
}

File: kalaidoscope/internal/api/organize.go
package api

type OrganizeStatus struct {
	Fragments int             `json:"fragments"`
	Imports   ImportsStatus   `json:"imports"`
	Map       MapStatus       `json:"map"`
	Discover  DiscoverStatus  `json:"discover"`
	Policy    OrganizePolicy  `json:"policy"`
	Reconcile ReconcileStatus `json:"reconcile"`
}

type ImportsStatus struct {
	Pending   int    `json:"pending"`
	LastError string `json:"lastError,omitempty"`
}

type MapStatus struct {
	State             string   `json:"state"`
	Version           int      `json:"version"`
	Annotated         int      `json:"annotated"`
	PendingAnnotation int      `json:"pendingAnnotation"`
	Unconsolidated    int      `json:"unconsolidated"`
	LastRun           *RunInfo `json:"lastRun,omitempty"`
	LastDrainError    string   `json:"lastDrainError,omitempty"`
}

type DiscoverStatus struct {
	State     string             `json:"state"`
	Running   string             `json:"running,omitempty"`
	Pending   []string           `json:"pending"`
	Due       []string           `json:"due"`
	Runs      map[string]RunInfo `json:"runs"`
	Proposals ProposalCounts     `json:"proposals"`
}

type ProposalCounts struct {
	Projections int `json:"projections"`
	Reflections int `json:"reflections"`
}

type OrganizePolicy struct {
	Wave bool `json:"wave"`
}

// ReconcileStatus is the speculative wave's state: whether one is running,
// when the latest one began, what ended the last one, and when one last ran
// clean. The dashboard's Start reads it to tell "still generating" from
// "the wave I started has ended" (LastStarted at or after the press, and not
// Running).
type ReconcileStatus struct {
	Running bool `json:"running"`
	// RFC3339; empty until a wave has started.
	LastStarted string `json:"lastStarted,omitempty"`
	LastError   string `json:"lastError,omitempty"`
	// RFC3339; empty until a wave has completed without error.
	LastCompleted string `json:"lastCompleted,omitempty"`
}

type RunInfo struct {
	ID          string `json:"id"`
	Status      string `json:"status"`
	Error       string `json:"error,omitempty"`
	Model       string `json:"model,omitempty"`
	Rounds      int    `json:"rounds,omitempty"`
	MapVersion  int    `json:"mapVersion,omitempty"`
	Finished    string `json:"finished"`
	Interrupted bool   `json:"interrupted,omitempty"`
}

const (
	MapStateEmpty         = "empty"
	MapStateUnannotated   = "unannotated"
	MapStateAnnotating    = "annotating"
	MapStateConsolidating = "consolidating"
	MapStateFolding       = "folding"
	MapStateSettled       = "settled"

	DiscoverStateNeverRun = "never_run"
	DiscoverStatePending  = "pending"
	DiscoverStateRunning  = "running"
	DiscoverStateDue      = "due"
	DiscoverStateSettled  = "settled"
)
